---
title: "CHiCAGO Vignette"
VignetteIndexEntry: 
output:
  BiocStyle::html_document:
    toc: true
---
<!--
%\VignetteEngine{knitr::rmarkdown}
%\VignetteIndexEntry{CHiCAGO Vignette}
%\VignetteDepends{}
%\VignetteKeywords{Chicago}
%\VignettePackage{Chicago}
-->
```{r style, echo = FALSE, results = 'asis'}
BiocStyle::markdown()
```

## Introduction

This vignette will walk you through a typical CHiCAGO analysis.

A wrapper to perform the analysis, called *runChicago.R*, can be found on our website: [FIXME URL].

WARNING: The data set used in this tutorial comes from the package *PCHiCdata*. It is only a small part of a real data set. Do not use any of the input data in your own analysis. We are going to use data from mouse embryonic stem cells (mESCs):

```{r}
library(PCHiCdata)
```

## Input files required

Before you start, you will need: 

1. Five restriction map information files:

- Restriction map file (.rmap) - a bed file containing coordinates of the restriction fragments. By default, 4 columns: chr, start, end, fragmentID.
- Bait map file (.baitmap) - a bed file containing coordinates of the baited restriction fragments, and their associated annotations. By default, 5 columns: chr, start, end, fragmentID, geneName.
- *nperbin* file (.npb), *nbaitsperbin* file (.nbpb), proxOEfile (.poe) - Precompute these tables from the .rmap and .baitmap files, using the Python scripts *countNperBin.py*, *countNbaitsPerBin.py* and *getProxOE.py*. 

It is easiest to put all of these files in the same directory. If you are using a standard restriction enzyme and bait map, you can use the files on our website [FIXME URL], otherwise you will need to generate them yourself using CHiCTools, also available on our website.

Examples of these files are also provided in the package, as follows. Do not use these files in your analysis!

```{r}
dataPath <- system.file("extdata", package="PCHiCdata")
testDesignDir <- file.path(dataPath, "mm9TestDesign")
dir(testDesignDir)
```

2. You will also need your input data files. These files can be obtained by running bam2chicago.sh, available from our website [FIXME URL]. Normally, the names of these input files end with *.chinput*.

```{r}
testDataPath <- file.path(dataPath, "mESCchinputFiles")
dir(testDataPath)

files <- c(
    file.path(testDataPath, "mESCrep1.chinput"),
    file.path(testDataPath, "mESCrep2.chinput")
  )
```

OPTIONAL: Note that, because this data set is smaller than usual, we need to input some custom settings; normally, you will not have to do this.
```{r}
settingsFile <- file.path(system.file("extdata", package="PCHiCdata"), "smESCSettings", "smESC.settingsFile")
```

## Example workflow

We run CHiCAGO on the test data as follows. First, we create a blank ``ChicagoData`` object, and we tell it where the design files are. For this example, we also provide the optional settings file.

```{r, message=FALSE}
library(Chicago)

cd <- setExperiment(designDir = testDesignDir, settingsFile = settingsFile)
```

The properties of ``ChicagoData`` objects are discussed more in [The ChicagoData object](#link).

Next, we read in the input data files:

```{r, message=FALSE}
cd <- readAndMerge(files=files, cd=cd)
```

Finally, we run the pipeline with ``chicagoPipeline()``:

```{r, eval=FALSE}
cd <- chicagoPipeline(cd)
```

## Output plots

``chicagoPipeline()`` produces a number of plots. You can save these to disk by setting the ``outprefix`` argument in ``chicagoPipeline()``.

The plots are as follows:

1. FIXME
2. FIXME
3. FIXME

```{r, echo=FALSE, message=FALSE}
cd <- chicagoPipeline(cd)
```

## Output files

You can export the results to disk, using ``exportResults()``. (If you use *runChicago.R*, the files appear in ./\<results-folder\>/data.) By default, the function outputs three different output file formats:

```{r}
outputDirectory <- tempdir()
exportResults(cd, file.path(outputDirectory, "vignetteOutput"))
```

Each called interaction is assigned a score that represents how strong CHiCAGO believes the interaction is: formally, it is -log(adjusted P-value). Thus, a larger score represents a stronger interaction. In each case, the score threshold of -log(adjusted p-value) of 5 is applied.

FIXME explain what happens for bait-to-bait interactions.

Summary of output files:

ibed format (ends with ...ibed):

- each row represents an interaction
- first four columns give information about the chromosome, start, end and name of the bait fragment
- next four columns give information about the chromosome, start, end and name of the other end that interacts with the bait fragment
- N\_reads is the number of reads
- score is as defined above

seqmonk format (ends with ...seqmonk.txt)

- Can be read by [seqmonk](http://www.bioinformatics.babraham.ac.uk/projects/seqmonk/).
- An interaction is represented by two rows: the first row is the bait, the second the other end. Thus, the file alternates: bait1, otherEnd1, bait2, otherEnd2, ...
- Columns are: chromosome, start, end, name, number of reads, interaction score (see above)

washU\_text format (ends with ...washU\_text.txt)

- Can be read by the [WashU browser](http://epigenomegateway.wustl.edu)
- Upload via the "Got text files instead? Upload them from your computer" link.
- Note - Advanced users may wish to export to washU\_track format instead. See the help page for ``exportResults()``.


## Peak enrichment for features

``peakEnrichment4Features()`` tests the hypothesis that other ends in the CHiCAGO output are enriched for genomic features of interest - for example, histone marks associated with enhancers. You will need additional files to perform this analysis - namely, a .bed file for each feature:

```{r}
#FIXME
```

## The ChicagoData object {#cd}

In the above workflow, *cd* is a *ChicagoData* object. It contains three elements:

* ``cd@x`` is a `r CRANpkg("data.table")` (note: not a *data.frame*) that contains information about fragment pairs.
* ``cd@settings`` is a list of settings, usually set with the setExperiment() function.
* ``cd@params`` is a list of parameters. This list is populated as the pipeline runs, and CHiCAGO estimates them in turn.

A closer look at ``cd@x``:
```{r}
head(cd@x, 2)
```

Columns:

* baitID: ID of baited fragment
* otherEndID: ID of other end fragment
* s_j: bait-specific scaling factor (Brownian noise)
* otherEndLen: The length of the other end fragment
* distSign: The distance from the baited fragment to the other end fragment. Positive and negative values indicate that the other end is, respectively, $5'$ and $3'$ of the baited fragment. NA indicates a trans interactions.
* isBait2Bait: TRUE if the other end fragment is also a baited fragment
* N.1, N.2: Normalized counts (see ``?mergeSamples``).
* N: Merged count (see ``?mergeSamples``).
* refBinMean: Can be ignored. (see ``?normaliseBaits``)
* s_i: other end-specific scaling factor (Brownian noise)
* NNb: "N normalised for baits", a count scaled up by accounting for s_j. May be useful for visualization.
* NNboe: "N normalised for baits and other ends"; may be useful for visualization.
* tlb: Class of other end, based on the number of fragments on other chromosomes that have read pairs.
* tblb: As tlb, for the bait fragment.
* Tmean: Expected count from technical noise.
* Bmean: Expected count from Brownian noise. (Thus, the expected count under the null hypothesis is Tmean + Bmean.)
* log.p: p-value associated with fragment pair, on log-scale.
* log.w: p-value weight, on log-scale.
* log.q: weighted p-value, on log-scale.
* score: Final CHiCAGO score.

**WARNING:** Many functions in CHiCAGO update ``cd@x`` by reference, which means that ``cd@x`` can change even when you do not explicitly assign to it.
To avoid this behaviour, copy the *chicagoData* object first:

```{r}
cdCopy <- cd
cdCopy@x <- copy(cd@x)
```

We will provide an explicit function to do this in due course.

##Session info

```{r}
sessionInfo()
```
